File: golang/two_sum_iv_input_is_a_bst.go
package main

import "golang/model"

func findTarget(root *model.TreeNode, k int) bool {
	if root == nil {
		return false
	}

	d := make(map[int]int)
	var q []*model.TreeNode
	q = append(q, root)

	for len(q) > 0 {
		node := q[0]
		q = q[1:]

		if _, ok := d[k-node.Val]; ok {
			return true
		}
		d[node.Val] = 1
		if node.Left != nil {
			q = append(q, node.Left)
		}
		if node.Right != nil {
			q = append(q, node.Right)
		}
	}

	return false
}

File: python3/two_sum_iv_input_is_a_bst.py
from .model import TreeNode

# Space : O(n)
# Time  : O(n)


class Solution:
    def findTarget(self, root: TreeNode, k: int) -> bool:
        if not root:
            return False
        s = set()
        q = [root]

        while q:
            node = q.pop()

            if k - node.val in s:
                return True
            s.add(node.val)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

        return False
